airflow/models/dataset.py

```diff
@@ -199,3 +199,86 @@ def __repr__(self):
         for attr in [x.name for x in self.__mapper__.primary_key]:
             args.append(f"{attr}={getattr(self, attr)!r}")
         return f"{self.__class__.__name__}({', '.join(args)})"
+
+
+class DatasetEvent(Base):
+    """
+    A table to store datasets events.
+
+    :param dataset_id: reference to Dataset record
+    :param extra: JSON field for arbitrary extra info
+    :param source_task_id: the task_id of the TI which updated the dataset
+    :param source_dag_id: the dag_id of the TI which updated the dataset
+    :param source_run_id: the run_id of the TI which updated the dataset
+    :param source_map_index: the map_index of the TI which updated the dataset
+
+    We use relationships instead of foreign keys so that dataset events are not deleted even
+    if the foreign key object is.
+    """
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    dataset_id = Column(Integer, nullable=False)
+    extra = Column(ExtendedJSON, nullable=True)
+    source_task_id = Column(StringID(), nullable=True)
+    source_dag_id = Column(StringID(), nullable=True)
+    source_run_id = Column(StringID(), nullable=True)
+    source_map_index = Column(Integer, nullable=True, server_default=text("-1"))
```

Comment:
Any reason why this has to have `source_` prefix? I feel it's good without the prefix

Reply:
Yeah, it's a good question.  So with dataset events, there tend to be multiple things implicated.  There's the task that updated the dataset.  But then there's the also dag or task that depends on the dataset, and will be triggered by the the update.  And e.g. with the queue table, we definitely needed to make clear we're talking about the target dag id.  Here though, I think you're probably right, we probably don't need it because it's only about writes.

Reply:
And I guess, the idea is just to make it clear, when looking at the dataset event, and when we see the references to a dag_id, for instance, it's clear we're referring to "the dag that wrote to the dataset".

E.g. when you are in the downstream TI that is processing the dataset update, and you are looking at this dataset event object (e.g. to see what was done to the dataset), and you see dag or task references on it -- will it be clear enough that those references mean "this is the dag / task that updated the dataset". 

Probably so but let me know what you think.

Reply:
Source would be upstream. Do we plan to record downstream events for when a dataset triggers a dag run?

Reply:
yeah like it's clarifying that this TI is the source of the event -- maybe in future we have events that have a specific target? dunno

Reply:
this all looks good to me for source events. Some sort of target event will be really helpful to visualize everything.

Reply:
Yeah need to think through best way to model this so that everything can be tied together.  We can discuss next week.